File: cmd/producer/multi.py
from kafka import KafkaProducer, KafkaAdminClient
from kafka.admin import NewTopic
from kafka.partitioner.default import DefaultPartitioner  # Updated import
import json
import random
import time
import signal
import threading
from datetime import datetime
from typing import Dict, Any

class Message:
    def __init__(self, id: str, category: str, value: float):
        self.id = id
        self.timestamp = datetime.now().isoformat()
        self.value = value
        self.category = category

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "timestamp": self.timestamp,
            "value": self.value,
            "category": self.category
        }

class CustomPartitioner(DefaultPartitioner):  # Inherit from DefaultPartitioner
    def __init__(self, partitions):
        super().__init__(partitions)
        self.category_partition_map = {
            "sales": 0,
            "traffic": 1,
            "users": 2,
            "errors": 3,
            "latency": 4
        }

    def __call__(self, key_bytes, all_partitions, available_partitions):
        if key_bytes is None:
            return super().__call__(key_bytes, all_partitions, available_partitions)
        
        category = key_bytes.decode('utf-8')
        return self.category_partition_map.get(category, super().__call__(key_bytes, all_partitions, available_partitions))

def create_producer() -> KafkaProducer:
    return KafkaProducer(
        bootstrap_servers=['localhost:9092'],
        value_serializer=lambda v: json.dumps(v).encode('utf-8'),
        key_serializer=lambda v: v.encode('utf-8'),
        partitioner=CustomPartitioner  # Use the custom partitioner
    )

def ensure_topic_exists(topic_name: str, num_partitions: int):
    try:
        admin_client = KafkaAdminClient(bootstrap_servers=['localhost:9092'])
        if topic_name not in admin_client.list_topics():
            topic = NewTopic(
                name=topic_name,
                num_partitions=num_partitions,
                replication_factor=1
            )
            admin_client.create_topics([topic])
            print(f"Created topic {topic_name} with {num_partitions} partitions")
        admin_client.close()
    except Exception as e:
        print(f"Error ensuring topic exists: {e}")

def produce_messages(producer: KafkaProducer, category: str, stop_event: threading.Event):
    message_count = 0
    
    while not stop_event.is_set():
        try:
            msg = Message(
                id=f"msg-{category}-{message_count}",
                category=category,
                value=random.random() * 100
            )

            future = producer.send(
                'analytics-topic',
                key=category,
                value=msg.to_dict()
            )
            
            if message_count % 100 == 0:
                metadata = future.get(timeout=1)
                print(f"Producer (Category: {category}): Message sent! "
                      f"Partition: {metadata.partition}, Offset: {metadata.offset}")
            
            message_count += 1
            time.sleep(0.1)  # 100ms delay between messages
            
        except Exception as e:
            print(f"Error producing message: {e}")
            
    print(f"Producer for category {category} shutting down...")

def main():
    # Ensure topic exists with correct number of partitions
    ensure_topic_exists('analytics-topic', 5)
    
    # Create producer
    producer = create_producer()
    
    # Setup shutdown handling
    stop_event = threading.Event()
    def signal_handler(signum, frame):
        print("\nShutdown signal received, closing producers...")
        stop_event.set()
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    # Start producer threads
    categories = ["sales", "traffic", "users", "errors", "latency"]
    threads = []
    
    for category in categories:
        thread = threading.Thread(
            target=produce_messages,
            args=(producer, category, stop_event)
        )
        thread.start()
        threads.append(thread)
    
    # Wait for all threads to complete
    for thread in threads:
        thread.join()
    
    # Clean up
    producer.close()

if __name__ == "__main__":
    main()

File: cmd/producer/multi.go
package main

import (
    "encoding/json"
    "fmt"
    "log"
    "math/rand"
    "os"
    "os/signal"
    "sync"
    "syscall"
    "time"

    "github.com/IBM/sarama"
)

// Message represents the data structure we'll send to Kafka
type Message struct {
    ID        string    `json:"id"`
    Timestamp time.Time `json:"timestamp"`
    Value     float64   `json:"value"`
    Category  string    `json:"category"`
}

// CustomPartitioner ensures messages from the same category go to the same partition
type CustomPartitioner struct {
    numPartitions int32
}

func (p *CustomPartitioner) Partition(message *sarama.ProducerMessage, numPartitions int32) (int32, error) {
    if p.numPartitions == 0 {
        p.numPartitions = numPartitions
    }

    // Get category from message key
    key, _ := message.Key.Encode()
    category := string(key)

    // Use consistent hashing to assign categories to partitions
    var partition int32
    switch category {
    case "sales":
        partition = 0
    case "traffic":
        partition = 1
    case "users":
        partition = 2
    case "errors":
        partition = 3
    case "latency":
        partition = 4
    default:
        partition = rand.Int31n(numPartitions)
    }

    // Ensure partition is within valid range
    return partition % numPartitions, nil
}

func (*CustomPartitioner) RequiresConsistency() bool {
    return true
}

func NewCustomPartitioner(topic string) sarama.Partitioner {
    return &CustomPartitioner{}
}

func main() {
    rand.Seed(time.Now().UnixNano())

    // Kafka configuration
    config := sarama.NewConfig()
    config.Producer.RequiredAcks = sarama.WaitForAll
    config.Producer.Retry.Max = 5
    config.Producer.Return.Successes = true
    config.Producer.Partitioner = NewCustomPartitioner

    // Create producer
    producer, err := sarama.NewSyncProducer([]string{"localhost:9092"}, config)
    if err != nil {
        log.Fatalf("Error creating producer: %v", err)
    }
    defer producer.Close()

    // Get metadata about the topic to verify number of partitions
    admin, err := sarama.NewClusterAdmin([]string{"localhost:9092"}, config)
    if err != nil {
        log.Fatalf("Error creating admin client: %v", err)
    }
    defer admin.Close()

    metadata, err := admin.DescribeTopics([]string{"analytics-topic"})
    if err != nil {
        log.Fatalf("Error describing topics: %v", err)
    }

    if len(metadata) == 0 {
        log.Fatal("No topic metadata received")
    }

    numPartitions := len(metadata[0].Partitions)
    log.Printf("Topic analytics-topic has %d partitions", numPartitions)

    // Channel for shutdown signaling
    signals := make(chan os.Signal, 1)
    signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

    // WaitGroup for goroutines
    var wg sync.WaitGroup
    done := make(chan bool)

    // Start producer goroutines
    numProducers := 5 // One producer per category
    for i := 0; i < numProducers; i++ {
        wg.Add(1)
        go func(producerID int) {
            defer wg.Done()
            produceMessages(producer, producerID, done)
        }(i)
    }

    // Wait for shutdown signal
    <-signals
    fmt.Println("\nShutdown signal received, closing producers...")
    close(done)
    wg.Wait()
}

func produceMessages(producer sarama.SyncProducer, producerID int, done chan bool) {
    categories := []string{"sales", "traffic", "users", "errors", "latency"}
    category := categories[producerID%len(categories)]
    
    ticker := time.NewTicker(100 * time.Millisecond)
    defer ticker.Stop()

    messageCount := 0

    for {
        select {
        case <-done:
            return
        case <-ticker.C:
            // Create message
            msg := Message{
                ID:        fmt.Sprintf("msg-%s-%d", category, messageCount),
                Timestamp: time.Now(),
                Value:     rand.Float64() * 100,
                Category:  category,
            }

            // Convert to JSON
            jsonData, err := json.Marshal(msg)
            if err != nil {
                log.Printf("Error marshaling message: %v", err)
                continue
            }

            // Create Kafka message
            kafkaMsg := &sarama.ProducerMessage{
                Topic: "analytics-topic",
                Value: sarama.StringEncoder(jsonData),
                Key:   sarama.StringEncoder(category),
            }

            // Send message
            partition, offset, err := producer.SendMessage(kafkaMsg)
            if err != nil {
                log.Printf("Error sending message: %v", err)
                continue
            }

            if messageCount%100 == 0 {
                log.Printf("Producer %d (Category: %s): Message sent! Partition: %d, Offset: %d\n",
                    producerID, category, partition, offset)
            }

            messageCount++
        }
    }
}